pkg/tbtc/node.go

```diff
@@ -205,3 +265,203 @@ func (n *node) joinDKGIfEligible(seed *big.Int, startBlockNumber uint64) {
+		// and more blocks than expected are mined in the meantime.
+		// Additionally, we want to strongly extend the delay period
+		// periodically in order to give some additional time for nodes to
+		// recover and re-fill their internal TSS pre-parameters pools.
+		if drl.attemptCounter > 1 {
+			delayBlocks := drl.delayBlocks
+			if drl.attemptCounter%drl.delayBlocksBumpFrequency == 0 {
+				delayBlocks *= drl.delayBlocksBumpFactor
+			}
+
+			drl.attemptStartBlock = drl.attemptStartBlock +
+				dkg.ProtocolBlocks() +
+				delayBlocks
+		}
+
+		// Exclude all members controlled by the operators that were not
+		// qualified for the current attempt.
+		excludedMembers := make([]group.MemberIndex, 0)
+		attemptSkipped := false
+		for i, operator := range drl.selectedOperators {
+			if !qualifiedOperatorsSet[operator] {
+				memberIndex := group.MemberIndex(i + 1)
+				excludedMembers = append(excludedMembers, memberIndex)
+
+				// If the given member was not qualified for the given attempt,
+				// mark this attempt as skipped in order to skip the execution
+				// and set up the next attempt properly.
+				if memberIndex == drl.memberIndex {
+					attemptSkipped = true
+				}
+			}
+		}
+
+		var result *dkg.Result
+		var attemptErr error
+
+		if !attemptSkipped {
+			result, attemptErr = dkgAttemptFn(&dkgAttemptParams{
+				index:           drl.attemptCounter,
+				startBlock:      drl.attemptStartBlock,
+				excludedMembers: excludedMembers,
+			})
+			if attemptErr != nil {
+				var imErr *dkg.InactiveMembersError
+				if errors.As(attemptErr, &imErr) {
+					for _, memberIndex := range imErr.InactiveMembersIndexes {
+						operator := drl.selectedOperators[memberIndex-1]
+						drl.inactiveOperatorsSet[operator] = true
+					}
+				}
+			}
+		}
+
+		if attemptSkipped || attemptErr != nil {
+			var err error
+			qualifiedOperatorsSet, err = drl.qualifiedOperatorsSet()
+			if err != nil {
+				return nil, fmt.Errorf(
+					"cannot get qualified operators for attempt [%v]: [%w]",
+					drl.attemptCounter+1,
+					err,
+				)
+			}
+
+			continue
+		}
+
+		return result, nil
+	}
+}
+
+// qualifiedOperatorsSet returns a set of operators qualified to participate
+// in the given DKG attempt.
+func (drl *dkgRetryLoop) qualifiedOperatorsSet() (map[chain.Address]bool, error) {
+	// If this is one of the first attempts and random retries were not started
+	// yet, check if there are known inactive operators. If the group quorum
+	// can be maintained, just exclude the members controlled by the inactive
+	// operators from the qualified set.
+	if drl.attemptCounter <= 5 &&
+		drl.randomRetryCounter == 0 &&
+		len(drl.inactiveOperatorsSet) > 0 {
+		qualifiedOperators := make(chain.Addresses, 0)
+		for _, operator := range drl.selectedOperators {
+			if !drl.inactiveOperatorsSet[operator] {
+				qualifiedOperators = append(qualifiedOperators, operator)
+			}
+		}
+
+		if len(qualifiedOperators) >= drl.chainConfig.GroupQuorum {
```

Comment:
This, in combination with the argument to `EvaluateRetryParticipantsForKeyGeneration` being `drl. selectedOperators`, means if *this* attempt pushes us below quorum, we go into random retries with the full group from sortition, rather than `drl.selectedOperators` right?

I think that's a good way to do it, but probably worth calling that out explicitly in a comment somewhere

Reply:
The `drl.selectedOperators` is exactly the sortition (selection) result but the rest is true. I posted a comment about that: 5aceaa60